sdk/src/azure/platform/az_pahoasync.c

```diff
@@ -0,0 +1,619 @@
+
+  _az_RETURN_IF_FAILED(
+      _az_result_from_pahoasync(MQTTAsync_connect(*me->_internal.pahoasync_handle, &conn_opts)));
+
+  return ret;
+}
+
+AZ_NODISCARD az_result az_mqtt5_outbound_sub(az_mqtt5* mqtt5, az_mqtt5_sub_data* sub_data)
+{
+  MQTTAsync_responseOptions opts = MQTTAsync_responseOptions_initializer;
+  opts.onSuccess5 = _az_pahoasync_on_subscribe_success;
+  opts.onFailure5 = _az_pahoasync_on_subscribe_failure;
+  opts.context = mqtt5;
+
+  return _az_result_from_pahoasync(MQTTAsync_subscribe(
+      *mqtt5->_internal.pahoasync_handle,
+      (const char*)az_span_ptr(sub_data->topic_filter),
+      sub_data->qos,
+      &opts));
+}
+
+AZ_NODISCARD az_result az_mqtt5_outbound_pub(az_mqtt5* mqtt5, az_mqtt5_pub_data* pub_data)
+{
+  MQTTAsync_responseOptions opts = MQTTAsync_responseOptions_initializer;
+  MQTTAsync_message msg = MQTTAsync_message_initializer;
+
+  opts.onSuccess5 = _az_pahoasync_on_publish_success;
+  opts.onFailure5 = _az_pahoasync_on_publish_failure;
+  opts.context = mqtt5;
+
+  msg.payload = (void*)az_span_ptr(pub_data->payload);
+  msg.payloadlen = az_span_size(pub_data->payload);
+  msg.qos = pub_data->qos;
+  if (pub_data->properties != NULL)
+  {
+    msg.properties = *(pub_data->properties->pahoasync_properties);
+  }
+
+  return _az_result_from_pahoasync(MQTTAsync_sendMessage(
+      *mqtt5->_internal.pahoasync_handle, (const char*)az_span_ptr(pub_data->topic), &msg, &opts));
+}
+
+AZ_NODISCARD az_result az_mqtt5_outbound_disconnect(az_mqtt5* mqtt5)
+{
+  MQTTAsync_disconnectOptions opts = MQTTAsync_disconnectOptions_initializer;
+
+  opts.onSuccess5 = _az_pahoasync_on_disconnect_success;
+  opts.onFailure5 = _az_pahoasync_on_disconnect_failure;
+  opts.context = mqtt5;
+
+  return _az_result_from_pahoasync(MQTTAsync_disconnect(*mqtt5->_internal.pahoasync_handle, &opts));
+}
+
+AZ_NODISCARD az_result az_mqtt5_property_bag_init(
+    az_mqtt5_property_bag* property_bag,
+    az_mqtt5* mqtt5,
+    MQTTProperties* pahoasync_properties)
+{
+  (void)mqtt5;
+  property_bag->pahoasync_properties = pahoasync_properties;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_mqtt5_property_bag_clear(az_mqtt5_property_bag* property_bag)
+{
+  _az_PRECONDITION_NOT_NULL(property_bag);
+
+  property_bag->pahoasync_properties->count = 0;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_mqtt5_property_bag_append_string(
+    az_mqtt5_property_bag* property_bag,
+    az_mqtt5_property_type type,
+    az_mqtt5_property_string* prop_str)
+{
+  _az_PRECONDITION_NOT_NULL(property_bag);
+  _az_PRECONDITION_NOT_NULL(prop_str);
+
+  property_bag->pahoasync_property.identifier = (enum MQTTPropertyCodes)type;
+  property_bag->pahoasync_property.value.data.data = (char*)az_span_ptr(prop_str->str);
+  property_bag->pahoasync_property.value.data.len = az_span_size(prop_str->str);
+
+  return _az_result_from_pahoasync(
+      MQTTProperties_add(property_bag->pahoasync_properties, &(property_bag->pahoasync_property)));
```

Comment:
Should the values set in pahoasync_property be cleared after they're added to the bag so you don't accidentally carry any values over later?

Reply:
I added a helper function to clear the property